### ISSUE TITLE
file descriptors cache needs to be disableable / configurable

### ISSUE THREAD

**Issue by kolayuk** [xssnick/tonutils-storage#31]
Cuz for now it could cause issues on fetching pieces on the other nodes like 
```
2024/09/11 13:51:50 [STORAGE] LOAD PIECE FROM 192.168.1.36:10501 ERR: failed to query piece 4. err: response deadline exceeded, err: context deadline exceeded
```
because node is waiting infinitely, also limit is too low for the lib as server usually have `unlimited` file descriptors setting.

But for nor now it is impossible to manage it, because 
[fs is private](https://github.com/xssnick/tonutils-storage/blob/master/storage/torrent.go#L125) (being used in torrent.getInternalPieces)
[limit is const](https://github.com/xssnick/tonutils-storage/blob/master/storage/fs.go#L10)

**Comment by Gealber** [comment]
@xssnick I'll take a look at it

**Comment by Gealber** [comment]
@kolayuk could you provide a fully reproducible scenario, the code snippet you provided is quite incomplete. Just the most simple way to reproduce it. Would be quite helpful.

**Comment by kolayuk** [comment]
Yes, if if would be possible to override fs without a limit / cache (custom implementation up to executable) - it would be a solution

**Comment by xssnick** [comment]
It is strange why it stuck and not downloads after some attempts, it blocks forever and never unlocks?

**Comment by kolayuk** [comment]
It will unlock probably at the point when FDs will go lower than 800, but it can take a lot of time if you're continue to serve all your uploads (or you have to manually stop some of them). But my point is this limitation makes some sense for consumer machines (as linux usually have 1024 default limitation) - but makes no sense to have it strict limitation in lib (not in cli) as lib can be used in server-side software and servers usually have 1M file descriptors limitation or
```
$ ulimit -n
unlimited
```
depends on kernel

**Comment by Gealber** [comment]
Yes, @kolayuk you are right the reason why its get locked is because the 800 limit is reached. And as long as `fs.Free()` is not called this counter of FDS won't decrease, hence not unlocking. About increasing the limit, I also agree on the point of at least making it configurable.

**Comment by Gealber** [comment]
> It is strange why it stuck and not downloads after some attempts, it blocks forever and never unlocks?

The reason is because the limit is reached, and the call to `Wait()` in `Acquire` method will block until the current FD counter decrease. At least this is what I think. Maybe we should think about how to make this limit configurable

**Comment by xssnick** [comment]
As an option, we can make it configurable by command line argument, with default value to unlimited (or something really big), then if someone wants to tune it he can use this argument to set a specific value

**Comment by kolayuk** [comment]
Actually PR above with the increase does not help, interfaces are needed to override that behaviour. Lock occurs from the mutexes, not that FD limit
